# python

In [29]:
import sys
sys.version

'3.10.9 | packaged by Anaconda, Inc. | (main, Mar  1 2023, 18:18:15) [MSC v.1916 64 bit (AMD64)]'

In [30]:
"welcome to fsds with chatgp @10am"

'welcome to fsds with chatgp @10am'

In [31]:
python idenifier = object=variable

SyntaxError: invalid syntax (1183132813.py, line 1)

syntax
= values
values == <identifier>

In [32]:
a = 5 #a - identifer & 5 is value

a = where a - identifier & 5 - value (int value)b = 5.6b- identifer & programming languange

In [33]:
b = 6
b

6

In [34]:
b = 7
b

7

In [35]:
a = 5
type(5)

int

In [36]:
b = 5.6
type(b)

float

In [37]:
c = "nit"
type(c)

str

In [38]:
d = True
type(d)

bool

In [39]:
e = False
type(e)

bool

In [40]:
a,b,c,d,e

(5, 5.6, 'nit', True, False)

In [41]:
id(A)

NameError: name 'A' is not defined

In [ ]:
print(id(a))
print(id(b))
print(id(c))
print(id(d))
print(id(e))

In [42]:
a = [10,20,30,40,50]
a

[10, 20, 30, 40, 50]

In [43]:
type(a)

list

In [44]:
len(a)

5

print| len | type | id

# 1 - python introduction (task-1) 




-: I hope everybody install Anconda software which i share to you guys right

-: Just wanted to know how many of know any programming language

-: If you dont know any programming language then you are the best person to learn PYTHON

-: python is very easy language

:-what is python?

Ans - python is highly recommanded programming language & object oriented language

:-Father of python - Guido van Rosam
Python came from fun tv show called ''complete monty python's flying circus'' - broadcasted in BBC channel
Python borrowed all concept from c,c++,java,unix ( so 
python is everything) thats why python very very powerfull tool

Python developed in NRI - (Netherland) & lot of people say that python is new language

Java released on 1995. python was released on 1989 officialy released on (feb 20th 1991)

It has a large and comprehensive standard library.

In [45]:
a = 5.7
type(a)

float

:-Now python is very popular based on software industry requirment because everybody wants to write very less code/concile code

:-Current market trend is - Machine learing, Artificial intelligence, data science & Iot(Internet of things)

:-which companies are used python - google,nasa,uber,netfliz,reddit,facebook, everywhere python used everywhere

:-python code can understand everybody & python is dynamic programming language

:-In python everything done by PVM (python virtual machine)
:-you can access python in any platform independent- windows, linux, mac one code can run in all the 4 platform & no need to write separate programe for every platform. Once you write code you can run in platform

:-Python is dynamically programming language (not required to declared data types)

:-Python is freeware and open source. Moving from one platform to other platform without changeing any code
Python contains rich libray - numpy,pandas so python is the best application for datascience

:-which scenario python can't be used - ( python can not perform in mobile application like android )

:-Flavours of python - cpython(C programming),jpython(java programming),Iron python(c#.net),Ruby python(Ruby based application programme),Anaconda

:-python(Bigdata,datascience)
Python 1.0 introduce in jan 1994 -- Noorganization is working now
:- Python 2.0 introduce in oct 2000 -- Noorganization is working now

:-Python 3.0 introduce in Dec 2008, 2016, 2017,---- latest version - 3.6, 3.6, 3.7, 3.8, 3.9, 3.10


In [46]:
# integer
a = 2
type(a)

int

In [47]:
# integer
b = 922337036
print(b)

922337036


In [48]:
# floating point
pi = 3.14
print(pi)

3.14


In [49]:
# string 
c = "A"
print(c)

A


In [50]:
# string
name = "john Doe"
print(name)


john Doe


In [51]:
#boolean
q = True
print(q)

True


In [52]:
#empty value or numll data type
x = None
print(x)

None


In [53]:


#Variable assignment works from left to right. So the following will give you an syntax error.
0 = x

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (399686651.py, line 2)

In [54]:
x = 0
x

0

:-There is a person whose name - Multiple names are to identify person.so finally the Name which can be used for identification purpose.

:-Name in the python programme is called IDENTIFIER (x = 10) (X - identifier)
*'''

:-Nameing ceremoney we have some rules to naming a child . e.g - Gods name,Ancestor Name,have to do some R & D. you cannot keep the child name as - Cat or dog right.. so parent have to follow some rule and keep their child naming ceremony.

:-*Rules to define Python Identifier & we will check those rules ==

<1 Alphabet (uppercae & lowercase) <2> Digits (0-9) # should not stat with digit <3> underscore(_)


In [55]:
cash123 = 10
cash123

10

In [56]:
123cash = 10
123cash

SyntaxError: invalid decimal literal (2567225020.py, line 1)

In [57]:
#x = 10 # x is the variable & 10 is thealue
cash = 10 # Identifier ruls alphabet
#ca$h = 10 # $ - symbole is not allowed in python identifier but in java is allowed
#ca$h
cash


10

In [58]:
ca$h = 20
ca$h

SyntaxError: invalid syntax (1112431759.py, line 1)

In [59]:
CASH = 20
#CASH
CASH

20

In [60]:
CASH1 = 30
#CASH1
CASH1

30

In [61]:
# <2> Identifiers should not starts with dight ====
sum123 = 20 # Digit rules identifier
#123total = 30
#123total
sum123

20

In [62]:
# <3> Identifiers are case sensitive
#total = 10 # Python are case sensitive
Abcde = 20
#total
Abcde

20

In [63]:
new = 30
NEW

NameError: name 'NEW' is not defined

In [64]:
Total5 = 30
Total

NameError: name 'Total' is not defined

In [65]:
def = 4.6
def

SyntaxError: invalid syntax (1737738290.py, line 1)

In [66]:
if = 780
if

SyntaxError: invalid syntax (1979711427.py, line 1)

In [67]:
IF = 780
IF

780

In [68]:
DEF = 5.6
DEF
#def is a key words

5.6

In [69]:
# <4> Keywords can not be assigned as identifier
#if = 10   # if is keyword
#DEF = 20  # def is keyword
for = 50
#DEF
#if
for


SyntaxError: invalid syntax (462765923.py, line 4)

In [70]:
FOR = 58
for

SyntaxError: invalid syntax (2510775588.py, line 2)

In [71]:
def = 30
def

SyntaxError: invalid syntax (2010559268.py, line 1)

In [72]:
if = 30
if

SyntaxError: invalid syntax (328627015.py, line 1)

In [73]:
IFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF = 56
IFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF

56

In [74]:
if = 60
if

SyntaxError: invalid syntax (2219948675.py, line 1)

In [75]:
# 5> NO length limit in python Identifier
wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwxxxxxxxx = 10
wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwxxxxxxxx


10

In [76]:
_abc_def_gef  = 20
_abc_def_gef 

20

:-Q & A for valid / Invalid identifier - 1>123AMX 2>Amx123
3>ml2ai 4>_abc_def_gef 5>def 6>else 7>ELSE


:----- RULES OF PYTHON IDENTIFIER ------ 1> A to Z, a to z, 0 - 9 2> Doesnot starts with digit 3> Case sensitive 4> Reserved words or keywords cannot be a identifier 5> Identifier cannot have a lenght limit 6> _ only allowed 7> NO special character is allowed


::========================= WE LEARNT ABOUT PYTHON IDENTIFIER IN DETAILS ============================

**35 RESERVED WORDS---

:-True, False, None ==> Represent Boolean data types

:-and, or, not, is ==> Represent the operators

:-if, else, elif ==> Represent the statement (# python switch,do..while statament is not available)

:-while, for, break, continue, return, in, yield ==> Represent the loop concept

:-try, except, finally, raise, assert ==> Represent for functionallity

::import,from,as,class,def,pass,global,nonlocal,lambda,del,with==>Represent the class,method,function


:-*NOTES -- 35 RESERVED WORDS ARE (ALPHABET) // *EXCEPT (True,False,None)

In [77]:
#a = True # hash is used for comment
a = True
a

True

In [78]:
#b = None
b = none
b

NameError: name 'none' is not defined

In [79]:
#c = False
c = false
c

NameError: name 'false' is not defined

In [80]:
# How to remembr all keywords --- (Interview questions)
# KEYWORD is the module run from IMPORT class
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

In [81]:
# write a coding to create index of these keywords --- IMP ---

import pandas as pd # pandas is the module to create a dataframe
df = pd.DataFrame(keyword.kwlist)
df

# Always remember python Index begins with '0'

# ================= RESERVED WORDS COMPLETED ===============================

0
0      False
1       None
2       True
3        and
4         as
5     assert
6      async
7      await
8      break
9      class
10  continue
11       def
12       del
13      elif
14      else
15    except
16   finally
17       for
18      from
19    global
20        if
21    import
22        in
23        is
24    lambda
25  nonlocal
26       not
27        or
28      pass
29     raise
30    return
31       try
32     while
33      with
34     yield

:-PYTHON DATA TYPES // (14) - INBUILD DATA TYPES -
1>int 2>float 3>complex 4>bool 5>str 6>bytes 7>bytearray DATA STRUCTURE ---> 8>range 9>list 10>tuple 11>set 12>frozenset 13>dict 14>None

:-python provides some inbuild function like -- <1> print() <2> type() <3> id()

:-int,float,complex,boolen is not represent object # Tricky question

:-except these 4 everythig object # Tricky question

NOTE - [**In python all 14 data types are object only]Thats why we called as python is object oriented program

In [82]:
"hello world"

'hello world'

In [83]:
# What is other inbuild datatype available except 14 datatypes-

a = 10
print(a)       # To find the variable 
#print(a)         # To find the value of variable
#type(a) # To find the data type
id(a)            # To find an address of an object

10


2780409692688

In [84]:
b = 10
id(b)

2780409692688

In [85]:
c = 20
id(c)

2780409693008

In [86]:
a = 10
b = 10
id(a)

2780409692688

In [87]:
a = 10
a
id(a)

2780409692688

int datatypes -

INT Datatypes - The No.without decimal point are called as INTEGRAL DATATYPES *int datatype how many ways represent values in 3ways -

2> Binary form --- (Base-2) -- (0,1) 3> Octal form --- (Base-8) -- (0,7)

In [88]:
a1 = 4809
#a
#print(a)
type(a1)

int

In [89]:
#2.Binary form(Base 2)
#a = 1111   # value is declared
b = 0b111 # Now pvm convert value to binary value
b
#a

7

In [90]:
b_1 = 0b11
b_1

3

In [91]:
b2 = 0b22
b2

SyntaxError: invalid digit '2' in binary literal (3245495910.py, line 1)

In [92]:


b1 = 111
b1

111

In [93]:
c = 0b1111
c

15

In [94]:
b3 = 0b222

SyntaxError: invalid digit '2' in binary literal (3363899208.py, line 1)

In [95]:
b = 0b11  # Now pvm convert value to binary value
b

3

In [96]:
#3. Octal form(Base 8)
#a = 111 # Value is declared
b1 = 0o111 # Now pvm covert value to octal value
b1
#a

73

In [97]:
# final summary of INTEGRAL DATATYPES 
a = 10
b = 0b10
c = 0o100
a
b
c

64

In [98]:

c1 = 0o33
c1

27

In [99]:
b


2

In [100]:
c

64

In [101]:

A = 78
type(A)

int


float datatypes -
employee sal - 5676.76diesel price - 67.25

:-These values are not integral value this is called as decimal value

:-Floating datatype you cannot declare Binary,Octal & Hexadecimal because python enterpretur not accept that

:-In our schools we learn about EXPONTIAL form -(1.2e3) this you can find in float datatypes & only letter 'e' can allowed

In [102]:
b = 67.9
b
type(b)

float

In [103]:
b1 = 0b1
b1

1

In [104]:
c = 0o11.6
c

SyntaxError: invalid syntax (3270742341.py, line 1)

In [105]:


# float datatypes - 
#b2 = 0b1111.22
#b2
#print(type(b))
#type(b2) 
#b
d = 0o4567.67 # This is octal
#e = 0b4567.89 # This is binary
d
#e
#b

SyntaxError: invalid syntax (2910200607.py, line 7)

In [106]:
f1 = 1E1
f1
#type(f1)

10.0

In [107]:
#f = 1e4# only 'e' letter is allowed
g = 2.4E3 # except 'E' you can't execute any programme
g
#f
#type(g)

2400.0

In [108]:
g1 = 23e3
g1

23000.0

In [109]:
e = 5.e3
#type(e)
e

5000.0

complex datatypes -

:-Complex datatype format are:-(a+bj) (a--Real part/b--Imaginary part/j^2=-1)

:-j is the compulsory value & there is no other value accepted in complex type

:-j^2 = -1

:-Value of j is (j square is equal to -1) (j =(square root of -1) is equal to (j^2 = -1) pure mathemetics so if you want to develop mathmetic application or scientific application then python is the best option

:-Real type any type base can be accepted but imaginary part allow only integer

In [110]:
x = 30+40j #assigned int value in real part & imaginary part
x
#type(x)

(30+40j)

In [111]:
type(x)


complex

In [112]:
y = 20.5+3j #assigned float value in real part & imaginary part
z = 30.8+20j #assigned float value in real part & real value in imaginary part
y + z
y + z
y*z
y/z

(0.5126646102740539-0.23549650017795704j)

In [113]:
#c = 15+0b111j # Imaginary part cannot be binary,octal
d = 0o11+15j # Real part can be binary,octal
#c
d

(9+15j)

In [114]:
d2 = 0b111+15j
d2

(7+15j)

In [115]:
e1 = 4 + 15a
e1

SyntaxError: invalid decimal literal (1489874699.py, line 1)

In [116]:
a1 = 20+30j
b1 = 40+50j
a1+b1
a1-b1
a1*b1
a1/b1


(0.5609756097560976+0.04878048780487805j)

In [117]:
a = 2+3j
type(a)

complex

In [118]:
a1 = 10+20j  # I want to know what the value of real part & imaginary part
a1.real # complex data type will use in mathmetic concept not that required for programming lang
#uage
#a1.imaginary
a1.imag

20.0

bool datatypes -

:-True/False - (only allowed boolean values)

:-False value -- 0 (internally memory level conversion happened)

:-True value -- 1
a = 10


In [119]:
a = 10
b = 20
c = a>b
c

False

In [120]:
True+True

2

In [121]:
True*True

1

In [122]:
True-True


0

In [123]:
True/True

1.0

In [124]:
False+False

0

In [125]:
False+True

1

In [126]:
True/True

1.0

In [127]:
True/False

ZeroDivisionError: division by zero

str datatypes -

:-enclosed in '' (single quote) // "" (double quote)

:-singa line we assined as '' // ""

:-multiline we assigned as ("' "')

:-single & double quotes are allowed only for single line

:-triple quotes are allowed for multi comments & also you can declare triple quotes in single line as well

In [128]:
naresh = '''good for datascience'''
naresh
#type(naresh)


'good for datascience'

In [129]:


naresh = '''good for datascience'''
#type(careerera)#'single quotes'
naresh
type(naresh)

str

In [130]:
naresh2 = '''good
            for datascience'''
#type(careerera)#' single quotes'
naresh2
#type(naresh)

'good\n            for datascience'

In [131]:
naresh = '''good for datascience'''
#type(careerera)#' single quotes'
naresh
#type(naresh)

'good for datascience'

In [132]:
naresh1 = '''good for 
 datascience'''
#type(naresh1
naresh1


'good for \n datascience'

In [133]:
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

In [134]:
a = '''hallo
how 
are 
you'''

a
#type(a) # SINLE & DOUBLE QUOTES ARE EQUAL 

'hallo\nhow \nare \nyou'

In [135]:
b = '''hello 
    hi'''
b

'hello \n    hi'

In [136]:
# single quote & double both are equl
# ''' multiline comment'''

In [137]:
b = '''('hallo' 
  'how'
    'are you')'''
    
    # ''' ''' ==> triple quotes are used for multiline comments
b
    # is for commenting 

"('hallo' \n  'how'\n    'are you')"

Type casting or Type conversion -

int() -- float() -- complex() -- bool() -- str()

In [138]:

# int(): - you can convert from other type to int type except complex 

int(10.123)  # float to int
#int(10+20j) # cannot convert from complex to int
int(True)     # Bool to int
int(False)   # Boll to int
int('10')    # string to int
int('ten') # amx is a character

ValueError: invalid literal for int() with base 10: 'ten'

In [139]:
# float(): -- convert from any type to float except complex

float(10)      # int to float
#float(10+20j)  # cannot convert complex to float
float(False)     # boolean to float
float('11')    # string to float

11.0

In [140]:
# complex(): -- covert any other type to complex type
# --- this only for 1 argument ------

complex(10)     # int to complex
complex(10.5)  # float to complex
complex(True)  # Bool to complex
complex(False) # Bool to complex
complex('10')  # string to complex 

# ----- Now we will check for 2 arguments ------
complex(10,20)   # int to complex
complex(10,20.5) # float to complex
#complex(10, 0b11j)
complex('10')    # string to complex, you cannot assign 2 argument

(10+0j)

In [141]:
a1 = 10+20j
type(a1)
a1.real
a1.imag

20.0

In [142]:
# Bool()  - (0 means false // 1 means non zero)

bool(0)      # int to bool
bool(-10)     # int to bool
bool(0.0)    # float to bool
bool(0.01)   # float to bool 
bool(10+20j) # complext to bool
bool(0+1j)   # complex to bool
bool(" ")      # string to bool(if argument is empty string then false)
bool('abc')# string to bool(if argument is not empty string then true)
bool(' ')    # space is also treated as character so non empty string

True

In [143]:
bool(-10)

True

In [144]:
bool(0+1j)

True

In [145]:
# str(): --- any type is possible in string 

str(10)     # int to string
str(10.50)  # float to string
str(True)   # bool to string
str(10+20j) # complex to string

'(10+20j)'

In [146]:
z1 = str(10)
z1

'10'

:-Fundamental Datatypes are which we covered so far & also we saw how to work on the type casting from one data type to other -

:-We cannot convert our complex data types to int and float

int() float() complex() bool() str()


:-Fundamental datatypes vs Immutability --


:- fundamental datatypes are immutable. what is immutable - once we crate the object we are not allow to perform any changes in that object . we can say that (non-changeable behaviour)

:-why immutability concept is required -- if you look at below exampl - how many object we created only 1 object which is 10 but how many reference we assinged -- 3 reference indicates to 1 object
:-bigest advantage of this approch is memory utilization & performance is also improved (pvm do not want to wsat memory)

:-you can create object with different name, but you cannot create object with same name

In [147]:
x2 = 10
y2 = 10
z2 = 20
print(id(x2))
print(id(y2))
print(id(z2))

2780409692688
2780409692688
2780409693008


:-Mutable -- Changeable-- once you create an object

:-Immutable -- Non-changeable

:-Fundamental data types are IMMUTABLE but (LIST is mutable)

:-Everything in python is an object
**This concept reusing same object such type of concept is define following ranges - 1> int ----> 0 to 256 2> bool ---> 

-----> Always 3> str ----> Always 4> float & complex ----> Can not performe the reusable concept

In [148]:
x = 10 #id - address of the memory location
y = 10
print(id(x))
print(id(y))

2780409692688
2780409692688


In [149]:
id(y)

2780409692688

In [150]:
# is operator
x = 20 # x,y = 20
y = 20
x is y
y is x

True

In [151]:

x = True
y = True
z = False
x is y
y is x
z is x
z is y

False

In [152]:
import keyword
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

In [153]:
import pandas as pd
import numpy as np

# data type
# Int --> variebel name  = value

In [154]:
a = 5
type(a)


int

In [155]:
1a = 

SyntaxError: invalid decimal literal (521520196.py, line 1)

In [156]:
if = 7

SyntaxError: invalid syntax (2329730394.py, line 1)

In [157]:
a1 = 1111
a1

1111

In [158]:
a2 = 0b11 
a2

3

In [159]:

a3 = 0b111
type(a3)

int

In [160]:




a4 = 11
a4

11

In [161]:
a5 = 0b2
a5

SyntaxError: invalid digit '2' in binary literal (461725866.py, line 1)

In [163]:



a6 = 0b3

SyntaxError: invalid digit '3' in binary literal (586578350.py, line 1)

In [164]:
a7 = 1b1

SyntaxError: invalid decimal literal (1441575157.py, line 1)

In [165]:
a = 0b101
a

5

In [166]:
bin(5)

'0b101'

In [167]:
bin(6)

'0b110'

In [168]:
0b110

6

In [169]:
0o110

72

In [170]:
0b110

6

In [171]:
0o111

73

In [174]:
oct(73)

'0o111'

In [175]:
0o76

62

In [176]:
0o9

SyntaxError: invalid digit '9' in octal literal (3061032682.py, line 1)

In [177]:
0o8

SyntaxError: invalid digit '8' in octal literal (1913375681.py, line 1)

In [178]:
0o7

7

In [179]:
09o

SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers (3303684915.py, line 1)

In [180]:
#float DataType

petrol = 110.89
petrol

110.89

In [181]:
type(petrol)

float

In [182]:
salary = 239888.45
salary

239888.45

In [183]:
petrol_ = 0b111.01

SyntaxError: invalid syntax (3752335488.py, line 1)

In [184]:
salary_ = 0o34567.78
salary_

SyntaxError: invalid syntax (1220410406.py, line 1)

In [185]:
f1 = 5e0
f1

5.0

In [186]:
type(f1)

float

In [187]:
5.3e1

53.0

In [188]:
5.3e2

530.0

In [189]:
5.3e3

5300.0

In [190]:
5.3E2

530.0

In [191]:
b = true

NameError: name 'true' is not defined

In [192]:
b = True
print(b)
print(type(b))

True
<class 'bool'>


In [193]:
keyword.kwlist

['False',
 'None',
 'True',
 'and',
 'as',
 'assert',
 'async',
 'await',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'nonlocal',
 'not',
 'or',
 'pass',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

In [194]:
True +True

2

In [195]:

True*True

1

In [196]:
True/False

ZeroDivisionError: division by zero

In [197]:

False/False

ZeroDivisionError: division by zero

In [198]:
False/True

0.0

In [199]:
False//True

0

In [200]:
bool(True)

True

In [201]:
int(True)

1

In [202]:
int(True) #type ca

1

In [203]:
a = True
a
type(a)

bool

In [204]:
int(a)

1

In [205]:
int(False)

0

In [206]:
c = 10 +20j
c

(10+20j)

In [207]:
type(c)

complex

In [208]:
c = 10 + 20j
c.real

10.0

In [209]:
c.imaginary

AttributeError: 'complex' object has no attribute 'imaginary'

In [210]:
c.imag

20.0

In [211]:
c

(10+20j)

In [212]:
print(c.real)

10.0


In [213]:
print(c.imag)

20.0


In [214]:
c_1 = 5 + 10j
c_2 = 15 + 20j

In [215]:
c_1 + c_2


(20+30j)

In [216]:
c_1 - c_2

(-10-10j)

In [217]:
c_1 * c_2

(-125+250j)

In [218]:
real
imaginary
j = squar root of -1

SyntaxError: invalid syntax (319071079.py, line 3)

In [219]:
# string


In [220]:
s = "welcome to nit"
s
type(s)

str

In [221]:
s1 = "welcom to nit"
s1
type(s1)

str

In [222]:
s1 = '''welcom
        to nit
            we are hear to learn ds , ai'''
s1

'welcom\n        to nit\n            we are hear to learn ds , ai'

In [223]:
a = 1 + 2j
b = 3 + 4j
a + b

(4+6j)

In [224]:
s1 = "welcome to fsds with chatgpt"
s1

'welcome to fsds with chatgpt'

In [225]:
type(s1)

str

In [226]:
s2 = "welcome to fsds with chatgpt"
s2

'welcome to fsds with chatgpt'

In [227]:


type(s2)

str

In [228]:
s3 = '''welcome to chatgpt'''
s3

'welcome to chatgpt'

In [229]:
s4 = '''welcome
        to
            fsds
                with chatgpt'''
s4

'welcome\n        to\n            fsds\n                with chatgpt'

In [230]:
chatgpt = '''We’ve trained a model called ChatGPT which interacts in a 
        conversational way. The dialogue format makes it possible for ChatGPT to answer 
        followup questions, admit its mistakes, challenge incorrect premises, and reject 
        inappropriate requests.'''

In [231]:
chatgpt

'We’ve trained a model called ChatGPT which interacts in a \n        conversational way. The dialogue format makes it possible for ChatGPT to answer \n        followup questions, admit its mistakes, challenge incorrect premises, and reject \n        inappropriate requests.'

In [232]:
type(chatgpt)

str

In [233]:
s5 = 10
type(s5)

int

In [234]:
s6 = "10"
s6

'10'

# today we are compled datat type
# DATA TYPE CONVERSION | TYPE CASTING
:- how i convert 1 data type other datatype argument = parameter german(bmw) == bmw - only 1 argument pass

In [235]:
int(3.4) # float to int conversion

3

In [236]:
int(3.4, 6.7)

TypeError: 'float' object cannot be interpreted as an integer

In [237]:
int(3+ 4j ) # complex to int not possible to convert this

TypeError: int() argument must be a string, a bytes-like object or a real number, not 'complex'

In [238]:
int(True)

1

In [239]:
int(False)

0

In [240]:
z1 = int("10")
z1

10

In [241]:
int("ten")

ValueError: invalid literal for int() with base 10: 'ten'

In [242]:
float(23)

23.0

In [243]:
float(23,35)

TypeError: float expected at most 1 argument, got 2

In [244]:
float(20_3j)

TypeError: float() argument must be a string or a real number, not 'complex'

In [245]:
float(True)

1.0

In [246]:
float(False)

0.0

In [247]:
float("ten")

ValueError: could not convert string to float: 'ten'

In [248]:
float("10")

10.0

In [249]:
complex(10)

(10+0j)

In [250]:
complex(20,30)

(20+30j)

In [251]:
complex(10.23)

(10.23+0j)

In [252]:
complex(True, False)

(1+0j)

In [253]:
complex(False, True)

1j

In [254]:
complex("10")

(10+0j)

In [255]:
complex("20")

(20+0j)

In [256]:
str(12.78)

'12.78'

In [257]:
str(12.78,6.7)

TypeError: str() argument 'encoding' must be str, not float

In [258]:
str(10+20j)

'(10+20j)'

In [259]:
str(int(True))

'1'

In [260]:
s7 = str(True)
print(s7)
print(type(s7))

True
<class 'str'>


In [261]:
s8 = str(int(True))
print(s8)
print(type(s8))


1
<class 'str'>


In [262]:
import tkinter as tk

#creat the main window title
window = tk.Tk()

#set the window title
window.title("My Tkinter Window")

# set the window size
window.geometry("400x300")

#Add a lebel to the window
label = tk.Label(window, text="Hello, TKinter!")
label.pack()

#Start the Tkinter event loop
window.mainloop()


# data structure
 LIST
 
 TUPLE
 
 SET
 
 DICT
 
 RANGE

In [263]:
car = "audi"
print(car)
print(type(car))

audi
<class 'str'>


In [264]:
car1 = ["audi", "bmw", "volks",  "jagwar"]
car1

['audi', 'bmw', 'volks', 'jagwar']

In [265]:
type(car)

str

In [266]:
type(car1)

list

In [267]:

type(car)

str

In [268]:
a = 1 # data type
b = 1,2,3 # data structure


In [269]:

type(a)

int

In [270]:
type(b)

tuple

# list

In [271]:
l = []
l

[]

In [272]:
l.append[10]

TypeError: 'builtin_function_or_method' object is not subscriptable

In [273]:
l.append(10)
l.append(20)
l

[10, 20]

In [274]:
type(l)

list

In [275]:
l.append(10)
l.append(20)
l

[10, 20, 10, 20]

In [276]:
l.append(30, 40, 50, 60)

TypeError: list.append() takes exactly one argument (4 given)

In [277]:
l.insert(30,40)
l

[10, 20, 10, 20, 40]

In [278]:
l.insert(50,60,70)

TypeError: insert expected 2 arguments, got 3

In [279]:
l

[10, 20, 10, 20, 40]

In [280]:
del l

In [281]:
l = []

In [282]:
l.append(10) # passing int argument
l.append(7.8) # passing float argument
l.append(True) # bool argument
l.append(1+2j) # complex argument
l. append("hello") # string

In [283]:
l

[10, 7.8, True, (1+2j), 'hello']

In [284]:
l[0]

10

In [285]:
l[-1]

'hello'

In [286]:
l

[10, 7.8, True, (1+2j), 'hello']

In [287]:
l[-5]

10

In [288]:
l[5]

IndexError: list index out of range

In [289]:
l

[10, 7.8, True, (1+2j), 'hello']

In [290]:
l[:]

[10, 7.8, True, (1+2j), 'hello']

In [291]:
l[2]

True

In [292]:
l[:2]

[10, 7.8]

In [293]:
l

[10, 7.8, True, (1+2j), 'hello']

In [294]:
l[:4]

[10, 7.8, True, (1+2j)]

In [295]:
l[:6]

[10, 7.8, True, (1+2j), 'hello']

In [296]:
l[6:]

[]

In [297]:
l[0:4]

[10, 7.8, True, (1+2j)]

In [298]:
l[0:4:1]

[10, 7.8, True, (1+2j)]

In [299]:
l

[10, 7.8, True, (1+2j), 'hello']

In [300]:
l[0:4:2]

[10, True]

In [301]:
import jupyterthemes as jt
from jupyterthemes.stylefx import set_nb_theme

# Change the theme to Jupyter Notebook ['light', 'dark', 'chesterish', 'gruvboxd', 'gruvboxl', 'solarizedd', 'solarizedl']
set_nb_theme('cheste)

SyntaxError: unterminated string literal (detected at line 5) (28312042.py, line 5)

In [302]:
import jupyterthemes as jt

In [303]:
l

[10, 7.8, True, (1+2j), 'hello']

In [304]:
l[0]

10

In [305]:
l[2]

True

In [306]:
l[:2]

[10, 7.8]

In [307]:
l[0] = 20

In [308]:
l

[20, 7.8, True, (1+2j), 'hello']

In [309]:
l[0] = 20
l

[20, 7.8, True, (1+2j), 'hello']

In [310]:
l[3] = "nit"
l

[20, 7.8, True, 'nit', 'hello']

In [311]:
l.pop()

'hello'

In [312]:
l

[20, 7.8, True, 'nit']

In [313]:
l.append("nit")

In [314]:
l.append("a", 23)

TypeError: list.append() takes exactly one argument (2 given)

In [315]:
l.insert(5,23)
l

[20, 7.8, True, 'nit', 'nit', 23]

In [316]:
l.insert(1,23)
l

[20, 23, 7.8, True, 'nit', 'nit', 23]

In [317]:
l.insert(1,300)

In [318]:
l

[20, 300, 23, 7.8, True, 'nit', 'nit', 23]

In [319]:
l.insert(1,"abc")

In [320]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', 'nit', 23]

In [321]:
l.insert(11)
l

TypeError: insert expected 2 arguments, got 1

In [322]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', 'nit', 23]

In [323]:
l[:10]

[20, 'abc', 300, 23, 7.8, True, 'nit', 'nit', 23]

In [324]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', 'nit', 23]

In [325]:
l.insert(7,2 + 2j)
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), 'nit', 23]

In [326]:
l.insert(7,2 + 2j)
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [327]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [328]:
l1 = ["nit", 25,[50, 100],[150, 90]]
l1

['nit', 25, [50, 100], [150, 90]]

In [329]:
type(l1)

list

In [330]:
l2=[]

In [331]:
l2 = l1.copy()
l2

['nit', 25, [50, 100], [150, 90]]

In [332]:
l1 == l2

True

In [333]:
l2

['nit', 25, [50, 100], [150, 90]]

In [334]:
l2.remove("nit")
l2

[25, [50, 100], [150, 90]]

In [335]:
l1 == l2

False

In [336]:
l1 != l2

True

In [337]:
l1 == l2

False

In [338]:
l2

[25, [50, 100], [150, 90]]

In [339]:
l2[1]

[50, 100]

In [340]:
l2[2]

[150, 90]

In [341]:
l2[:4]

[25, [50, 100], [150, 90]]

In [342]:
l2

[25, [50, 100], [150, 90]]

In [343]:
l2[1]

[50, 100]

In [344]:
l2[2]

[150, 90]

In [345]:
l2[:4]

[25, [50, 100], [150, 90]]

In [346]:
l2[4:]

[]

In [347]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [348]:
l1

['nit', 25, [50, 100], [150, 90]]

In [349]:
l2

[25, [50, 100], [150, 90]]

In [350]:
del l2

In [351]:
l2

NameError: name 'l2' is not defined

In [352]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [353]:
l1

['nit', 25, [50, 100], [150, 90]]

In [354]:
print(len(l))
print(len(l1))

11
4


In [355]:
len(l1)

4

In [356]:
len(l)

11

In [357]:
print(id(l1))
print(id(l))

2780560933312
2780560674112


In [358]:
l2 = l1.copy()

In [359]:
print(id(l))
print(id(l1))
print(id(l2))

2780560674112
2780560933312
2780560860032


In [361]:
print(id(l))
print(id(l1))
print(id(l2))

2780560674112
2780560933312
2780560860032


In [362]:
print(l)
print(len(l))
#===========
print(l1)
print(len(l1))
#==============
print(l2)
print(len(l2))

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]
11
['nit', 25, [50, 100], [150, 90]]
4
['nit', 25, [50, 100], [150, 90]]
4


In [364]:
a1 = 5
b1 = 6
c1 = 6
print(id(a1))
print(id(b1))
print(id(c1))

2780409692528
2780409692560
2780409692560


In [365]:
l2

['nit', 25, [50, 100], [150, 90]]

In [366]:
l2[4] = {2, 3 , 4}

IndexError: list assignment index out of range

In [367]:
l2.insert(4, {2, 3, 4})
l2

['nit', 25, [50, 100], [150, 90], {2, 3, 4}]

In [368]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [369]:
l2

['nit', 25, [50, 100], [150, 90], {2, 3, 4}]

In [370]:
l2.add(5)

AttributeError: 'list' object has no attribute 'add'

In [371]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [372]:
l.reverse()

In [373]:
l

[23, 'nit', (2+2j), (2+2j), 'nit', True, 7.8, 23, 300, 'abc', 20]

In [374]:
l2.reverse()

In [375]:
l2

[{2, 3, 4}, [150, 90], [50, 100], 25, 'nit']

In [376]:
l1

['nit', 25, [50, 100], [150, 90]]

In [377]:
l2

[{2, 3, 4}, [150, 90], [50, 100], 25, 'nit']

In [378]:
l3 = l.reverse()
l4 = l1.reverse()
l5 = l2.reverse()

In [379]:
print(l)
print(l1)
print(l2)
print(l3)
print(l4)
print(l5)

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]
[[150, 90], [50, 100], 25, 'nit']
['nit', 25, [50, 100], [150, 90], {2, 3, 4}]
None
None
None


In [380]:
l1 = l1.reverse()

In [381]:
l4 = [20,"abc", 300, 7.8,True, "nit",23, (2+2j)]
l4

[20, 'abc', 300, 7.8, True, 'nit', 23, (2+2j)]

In [382]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [383]:
l2

['nit', 25, [50, 100], [150, 90], {2, 3, 4}]

In [385]:
b = l[::-1]
b

[23, 'nit', (2+2j), (2+2j), 'nit', True, 7.8, 23, 300, 'abc', 20]

In [386]:
 l_1 = [1,2,3]
 l_1   

[1, 2, 3]

In [388]:
l_1.reverse()
l_1

[3, 2, 1]

In [389]:
l_2 = l_1[::-1]

In [390]:
l_2

[1, 2, 3]

In [391]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [394]:
l[6]

'nit'

In [395]:
print(l[6][0])
print(l[6][1])
print(l[6][2])

n
i
t


In [396]:
l

[20, 'abc', 300, 23, 7.8, True, 'nit', (2+2j), (2+2j), 'nit', 23]

In [408]:
l[6][1]

'i'

In [411]:
l.remove(True)
l

[20, 'abc', 300, 23, 7.8, 'nit', (2+2j), (2+2j), 'nit', 23]

add,remove,change items in  the list

In [413]:
l

[20, 'abc', 300, 23, 7.8, 'nit', (2+2j), (2+2j), 'nit', 23]

In [414]:
l2

['nit', 25, [50, 100], [150, 90], {2, 3, 4}]

In [416]:
del l4

In [417]:
l

[20, 'abc', 300, 23, 7.8, 'nit', (2+2j), (2+2j), 'nit', 23]

In [418]:
l2

['nit', 25, [50, 100], [150, 90], {2, 3, 4}]

In [419]:
l_1

[3, 2, 1]

In [420]:
b

[23, 'nit', (2+2j), (2+2j), 'nit', True, 7.8, 23, 300, 'abc', 20]

In [422]:
b.remove("nit")
b

[23, (2+2j), (2+2j), True, 7.8, 23, 300, 'abc', 20]

In [428]:
l

[20, 'abc', 300, 23, 7.8, 'nit', (2+2j), (2+2j), 'nit', 23]

In [429]:
l2

['nit', 25, [50, 100], [150, 90], {2, 3, 4}]

In [430]:
l

[20, 'abc', 300, 23, 7.8, 'nit', (2+2j), (2+2j), 'nit', 23]

In [431]:
l[3:5]

[23, 7.8]

In [433]:
del l[3:5]
l

[20, 'abc', 300, (2+2j), 'nit', 23]

In [434]:
l

[20, 'abc', 300, (2+2j), 'nit', 23]

In [435]:
l2.clear()

In [436]:
l2

[]

In [437]:
l

[20, 'abc', 300, (2+2j), 'nit', 23]

In [438]:
20 in l

True

In [439]:
20,300 in l

(20, True)

In [440]:
l

[20, 'abc', 300, (2+2j), 'nit', 23]

In [441]:
l.sort()

TypeError: '<' not supported between instances of 'str' and 'int'

In [443]:
mylist3 = [6,3,7,0,6,9,3,4,8,5]
mylist3.sort(reverse = False)

In [444]:
mylist3

[0, 3, 3, 4, 5, 6, 6, 7, 8, 9]

In [445]:
mylist3

[0, 3, 3, 4, 5, 6, 6, 7, 8, 9]

In [446]:
mylist3.sort(reverse = True)
mylist3

[9, 8, 7, 6, 6, 5, 4, 3, 3, 0]